packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs

```diff
@@ -6,81 +6,62 @@ use postgres_types::ToSql;
+        AsClient, PostgresStore, QueryError,
+    },
 };
 
 #[async_trait]
-impl<C: AsClient> crud::Read<'_, EntityId, PersistedEntity> for PostgresStore<C> {
-    type Output = PersistedEntity;
+impl<C: AsClient> crud::Read<PersistedEntity> for PostgresStore<C> {
+    type Query<'q> = EntityQuery;
 
-    async fn get(&self, identifier: EntityId) -> Result<Self::Output, QueryError> {
-        let row = self
-            .as_client()
-            .query_one(
-                r#"
-                SELECT properties, entity_id, entities.version, ids.base_uri, ids.version, created_by
-                FROM entities
-                INNER JOIN ids ON ids.version_id = entities.entity_type_version_id
-                WHERE entity_id = $1 AND entities.version = (
-                    SELECT MAX("version")
-                    FROM entities
-                    WHERE entity_id = $1
-                );
-                "#,
-                &[&identifier],
-            )
-            .await
-            .into_report()
-            .change_context(QueryError)
-            .attach_printable(identifier)?;
-
-        let entity = serde_json::from_value(row.get(0))
-            .into_report()
-            .change_context(QueryError)?;
-
-        let entity_id: EntityId = row.get(1);
-        let entity_version: DateTime<Utc> = row.get(2);
-        let type_base_uri: String = row.get(3);
-        let type_version: i64 = row.get(4);
-        let created_by: AccountId = row.get(5);
-
-        let type_versioned_uri = VersionedUri::new(type_base_uri, type_version as u32);
-        Ok(PersistedEntity::new(
-            entity,
-            entity_id,
-            entity_version,
-            type_versioned_uri,
-            created_by,
-        ))
-    }
-}
-
-#[async_trait]
-impl<C: AsClient> crud::Read<'_, AllLatest, PersistedEntity> for PostgresStore<C> {
-    type Output = Vec<PersistedEntity>;
-
-    async fn get(&self, _: AllLatest) -> Result<Self::Output, QueryError> {
-        let row_stream = self
-            .as_client()
-            .query_raw(
-                r#"
-                SELECT DISTINCT ON(entity_id) properties, entity_id, entities.version, ids.base_uri, ids.version, created_by
-                FROM entities
-                INNER JOIN ids ON ids.version_id = entities.entity_type_version_id
-                ORDER BY entity_id, entities.version DESC;
-                "#,
-                // Requires a concrete type, which implements
-                // `IntoIterator<Item = impl BorrowToSql>`
-                [] as [&(dyn ToSql + Sync); 0],
-            )
-            .await
-            .into_report()
-            .change_context(QueryError)?;
+    async fn read<'query>(
+        &self,
+        query: &Self::Query<'query>,
+    ) -> Result<Vec<PersistedEntity>, QueryError> {
+        let row_stream = match (query.id(), query.version()) {
+            (Some(entity_id), Some(EntityVersion::Latest)) => {
```

Comment:
I'm a little worried about the maintainability of this, it seems less than ideal that we're explicitly managing permutations.

I'm fine with us merging as is and trying to keep an eye on it though and see how it develops as we add more constraints to the queries.

Reply:
I added a related comment here: https://github.com/hashintel/hash/pull/909#discussion_r942116961

I'm also not super happy with this currently, but for now, this is working fine. I will keep this in mind when further extending the queries.